### DISCUSSION TITLE
Database update 3->4->5?   or how to wipe the db and start over?

### DISCUSSION THREAD

**Discussion by sgtell** [ampache/ampache#3418] [Q&A]
Next issue in installation: I restored the database from my old installation of ampache ~3.7.  The 5.5.6 install process sees the db,
and  presents a page`http://ampache/update.php?action=update`
with

```
This page handles all database updates to Ampache starting with 3.3.3.5. Your current version is 5.5.6-release with database version 3.6 Build:0004

The following updates need to be performed:

    Version: 3.6 Build:0005
    * Modify tmp_browse to allow caching of multiple browses per session.
 ...
   lots more up thru
Version: 5.5 Build:0005
* Add `song_artist` and `album_artist` maps to catalog_map 

```

But pushing the "Update Now" button at the bottom takes me back to the very same page.  Is this supposed to work to step through all of the db schema updates?  Where would I look for error messages along the way?  
Or failing that, how do I flush the database and start over?  Just drop the database in mysql?  Then how to get the install to go back to create a new empty database?

**Comment by lachlan-00** [comment]
I've put all the database updates in there to allow updating from any major version so we should make sure that works.

The updates are all copied from the last 3.9 release so there is probably something missing, copied incorrectly OR your version is still missing something else before 3.6 Build:0005

The debug log will give you query errors which can tell me what's wrong. From there i can give you the queries to fix them or update the update file to correct any issues.

It is going to take forever for you most likely but the cli update will also help you go a bit faster.
https://github.com/ampache/ampache/wiki/ampache6-details#the-album-table-is-now-per-album-instead-of-per-disk

```
php bin/cli admin:updateDatabase -e
```

**first step** what is the actual DB version you have?

```
SELECT * FROM `update_info` WHERE `key` = 'db_version';
```

step 2 is what are the errors in the log. is it the create or the drop causing issues.

![image](https://user-images.githubusercontent.com/1305249/208322581-fe957dc3-5364-48a3-a857-05418e34248d.png)

**Reply by sgtell** [reply]
Thanks for the reply.
1.  
```
MariaDB [ampache]> SELECT * FROM `update_info` WHERE `key` = 'db_version';
+------------+--------+
| key        | value  |
+------------+--------+
| db_version | 360004 |
+------------+--------+
```

3. Following https://github.com/ampache/ampache/wiki/ampache6-details#the-album-table-is-now-per-album-instead-of-per-disk, I tried       `php bin/cli admin:updateDatabase` and `php bin/cli admin:updateDatabase -e` 
Heres the thing:  both with and without `-e` , that command runs really fast and produces only the long list of updates to be applied:  as if  the 'execute' option is ignored.   
I don't really know php, but I'm not afraid to hack in print statements :-) and see what happens.  So far I can see in src/Module/Cli/AdminUpdateDatabaseCommand.php  that the -e/--execute _is_ getting parsed correctly .  

I can see where  update_360005() is defined in src/Module/System/Update.php,  but I don't yet understand where it is supposed to get called from.

**Reply by lachlan-00** [reply]
Most updates should be quick in general. so you're on the step behind which is good.

That tells me that 360005 is failing. Are there errors in your log files?

The errors will look like this (Error_query SQL and Error_query MSG get printed to the log when there is a query error.)
```
2022-12-17T04:44:53+10:00 [ampache] (Ampache\Module\System\Dba) -> Error_query SQL: SELECT `song`.`id`, `song`.`file`, `song`.`catalog`, `song`.`album`, `song`.`total_count`, `song`.`total_skip`, `album`.`album_artist` AS `albumartist`, `song`.`year`, `song`.`artist`, `song`.`title`, `song`.`bitrate`, `song`.`rate`, `song`.`mode`, `song`.`size`, `song`.`time`, `song`.`track`, `song`.`played`, `song`.`enabled`, `song`.`update_time`, `song`.`mbid`, `song`.`addition_time`, `song`.`license`, `song`.`composer`, `song`.`user_upload`, `album`.`disk`, `album`.`mbid` AS `album_mbid`, `artist`.`mbid` AS `artist_mbid`, `album_artist`.`mbid` AS `albumartist_mbid` FROM `song` LEFT JOIN `album` ON `album`.`id` = `song`.`album` LEFT JOIN `artist` ON `artist`.`id` = `song`.`artist` LEFT JOIN `artist` AS `album_artist` ON `album_artist`.`id` = `album`.`album_artist` WHERE `song`.`id` = ? [787670]
2022-12-17T04:44:53+10:00 [ampache] (Ampache\Module\System\Dba) -> Error_query MSG: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'album.disk' in 'field list'
```

This update drops a table and recreates it. i think the issue is with the create sql

Do these queries to manually recreate the table work? (edited to force utf8)
```
DROP TABLE IF EXISTS `tmp_browse`;
CREATE TABLE `tmp_browse` (`id` int(13) NOT NULL auto_increment, `sid` varchar(128) CHARACTER SET utf8 NOT NULL default '', `data` longtext NOT NULL, `object_data` longtext, PRIMARY KEY  (`sid`, `id`)) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```

**Reply by sgtell** [reply]
Thanks... I'd just about gotten to that by studying the code (and adding more print statements).

The CREATE TABLE is failing when run manually in mysql:
```
MariaDB [ampache]> CREATE TABLE `tmp_browse` (`id` int(13) NOT NULL auto_increment, `sid` varchar(128) CHARACTER SET utf8 NOT NULL default '', `data` longtext NOT NULL, `object_data` longtext, PRIMARY KEY  (`sid`, `id`)) ENGINE=InnoDB DEFAULT CHARSET=utf8;
ERROR 1075 (42000): Incorrect table definition; there can be only one auto column and it must be defined as a key
MariaDB [ampache]> 
```

the "mysql" is mariadb-10.5.16-2.el9_0.x86_64

But where do i find ampache's logfiles?   Is there a subdirectory under the installation besides config that is supposed to be writable by the web server?   Or are they in a table in the db?   I did look for a cli command to dump them...

**Reply by lachlan-00** [reply]
That's what we want! I'll rework that query now.

That looks like sid and id as primary key are incorrect. it may be fixed later in the update so i'll fix that and post you some queries to manually apply the update.

the log files are configured in the config file and is printed to a local file. https://github.com/ampache/ampache/wiki/Basic#configuring-ampache

Usually you have to make that directory and chown to your web user. and uncomment the log parts in the config.

**Reply by lachlan-00** [reply]
thank you for being patient and going through these with me too

**Reply by lachlan-00** [reply]
can you try the query from the install file as well? I think it may be the default part of the sid column

```
DROP TABLE IF EXISTS `tmp_browse`;
CREATE TABLE IF NOT EXISTS `tmp_browse` (`id` int(13) NOT NULL AUTO_INCREMENT, `sid` varchar(128) COLLATE utf8_unicode_ci NOT NULL, `data` longtext COLLATE utf8_unicode_ci NOT NULL, `object_data` longtext COLLATE utf8_unicode_ci DEFAULT NULL, PRIMARY KEY (`sid`,`id`)) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```

**Reply by sgtell** [reply]
Those work fine:
```
MariaDB [ampache]> DROP TABLE IF EXISTS `tmp_browse`;
Query OK, 0 rows affected (0.000 sec)

MariaDB [ampache]> CREATE TABLE IF NOT EXISTS `tmp_browse` (`id` int(13) NOT NULL AUTO_INCREMENT, `sid` varchar(128) COLLATE utf8_unicode_ci NOT NULL, `data` longtext COLLATE utf8_unicode_ci NOT NULL, `object_data` longtext COLLATE utf8_unicode_ci DEFAULT NULL, PRIMARY KEY (`sid`,`id`)) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
Query OK, 0 rows affected (0.016 sec)
```

Should I patch that into Update.php and try pressing on?    
I realize that I still have the mysqldump, so this experimentation botches somthing up, I ought to be able to roll back and try again.

**Comment by sgtell** [comment]
altering  update_360005() lets it get all the way up through update_370017(), then 370018 fails.
I'm pretty sure its the first create table:
```
$sql = "CREATE TABLE IF NOT EXISTS `tag_merge` (`tag_id` int(11) NOT NULL, `merged_to` int(11) NOT NULL, FOREIGN KEY (`tag_id`) REFERENCES `tag` (`tag_id`), FOREIGN KEY (`merged_to`) REFERENCES `tag` (`tag_id`), PRIMARY KEY (`tag_id`, `merged_to`)) ENGINE=$engine";
```

**Reply by lachlan-00** [reply]
that's good, it's moving. gimme about 20 mins to finish up something and i'll look at that one.

**Reply by lachlan-00** [reply]
Lets add a drop in there and we'll recreate that table using the current sql file

```
    public static function update_370018(): bool
    {
        $retval  = true;
        $sql     = "DROP TABLE IF EXISTS `tag_merge`;";
        $retval &= (Dba::write($sql) !== false);
        $sql = "CREATE TABLE IF NOT EXISTS `tag_merge` (`tag_id` int(11) NOT NULL, `merged_to` int(11) NOT NULL, PRIMARY KEY (`tag_id`,`merged_to`), KEY `merged_to` (`merged_to`)) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;";
        $retval &= (Dba::write($sql) !== false);
        $sql = "INSERT INTO `tag_merge` (`tag_id`, `merged_to`) SELECT `tag`.`id`, `tag`.`merged_to` FROM `tag` WHERE `merged_to` IS NOT NULL";
        $retval &= (Dba::write($sql) !== false);
        $sql = "ALTER TABLE `tag` DROP COLUMN `merged_to`";
        $retval &= (Dba::write($sql) !== false);
        $sql = "ALTER TABLE `tag` ADD COLUMN `is_hidden` TINYINT(1) NOT NULL DEFAULT 0";
        $retval &= (Dba::write($sql) !== false);

        return $retval;
    }
```

**Reply by sgtell** [reply]
That's failing on the 'INSERT INTO...'; if I replay that interactively, I get this result.   Tried to diagnose, but I don't quite see it. 
```

MariaDB [ampache]> describe tag_merge;
+-----------+---------+------+-----+---------+-------+
| Field     | Type    | Null | Key | Default | Extra |
+-----------+---------+------+-----+---------+-------+
| tag_id    | int(11) | NO   | PRI | NULL    |       |
| merged_to | int(11) | NO   | PRI | NULL    |       |
+-----------+---------+------+-----+---------+-------+
2 rows in set (0.001 sec)

MariaDB [ampache]> INSERT INTO `tag_merge` (`tag_id`, `merged_to`) SELECT `tag`.`id`, `tag`.`merged_to` FROM `tag` WHERE `merged_to` IS NOT NULL;
ERROR 1054 (42S22): Unknown column 'tag.merged_to' in 'field list'
MariaDB [ampache]> describe tag;
+-----------+------------------+------+-----+---------+----------------+
| Field     | Type             | Null | Key | Default | Extra          |
+-----------+------------------+------+-----+---------+----------------+
| id        | int(11) unsigned | NO   | MUL | NULL    | auto_increment |
| name      | varchar(255)     | YES  | UNI | NULL    |                |
| is_hidden | tinyint(1)       | NO   |     | 0       |                |
+-----------+------------------+------+-----+---------+----------------+
3 rows in set (0.001 sec)

```
Perhaps somthing much earlier didn't get done?

**Reply by sgtell** [reply]
I'm willing to keep iterating like this for as long as you are; maybe I'll learn something :-)

**Reply by lachlan-00** [reply]
I really enjoy going back into this older part so we'll keep working through

The error means it's already dropped that column which it shouldn't do if it fails to insert But that's okay they're probably already been converted over for you.

Lets put a check in for the column first and then only delete after it's ok

For you this will skip those two queries but will continue
```
    /*
     * update 370018
     *
     * Enhance tag persistent merge reference.
     */
    public static function update_370018(): bool
    {
        $retval  = true;
        $sql     = "DROP TABLE IF EXISTS `tag_merge`;";
        $retval &= (Dba::write($sql) !== false);
        $sql = "CREATE TABLE IF NOT EXISTS `tag_merge` (`tag_id` int(11) NOT NULL, `merged_to` int(11) NOT NULL, PRIMARY KEY (`tag_id`,`merged_to`), KEY `merged_to` (`merged_to`)) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;";
        $retval &= (Dba::write($sql) !== false);
        $sql        = "DESCRIBE `tag`";
        $db_results = Dba::read($sql);
        while ($row = Dba::fetch_assoc($db_results)) {
            if ($row['Field'] == 'merged_to') {
                $sql = "INSERT INTO `tag_merge` (`tag_id`, `merged_to`) SELECT `tag`.`id`, `tag`.`merged_to` FROM `tag` WHERE `merged_to` IS NOT NULL";
                $retval &= (Dba::write($sql) !== false);
                if ($retval) {
                    // don't drop until you've confirmed a merge
                    $sql = "ALTER TABLE `tag` DROP COLUMN `merged_to`";
                    $retval &= (Dba::write($sql) !== false);
                }
            }
        }
        $sql = "ALTER TABLE `tag` ADD COLUMN `is_hidden` TINYINT(1) NOT NULL DEFAULT 0";
        $retval &= (Dba::write($sql) !== false);

        return $retval;
    }
```

**Reply by sgtell** [reply]
That choked on the final "ALTER TABLE `tag` ADD COLUMN `is_hidden`"  ;  somehow that column was already there (huh?)

```
MariaDB [ampache]> ALTER TABLE `tag` ADD COLUMN `is_hidden` TINYINT(1) NOT NULL DEFAULT 0;
ERROR 1060 (42S21): Duplicate column name 'is_hidden'
MariaDB [ampache]> describe tag;
+-----------+------------------+------+-----+---------+----------------+
| Field     | Type             | Null | Key | Default | Extra          |
+-----------+------------------+------+-----+---------+----------------+
| id        | int(11) unsigned | NO   | MUL | NULL    | auto_increment |
| name      | varchar(255)     | YES  | UNI | NULL    |                |
| is_hidden | tinyint(1)       | NO   |     | 0       |                |
+-----------+------------------+------+-----+---------+----------------+
3 rows in set (0.001 sec)
```
I moved ahead by hacking in a DROP COLUMN just before - but not sure if that's necessary or even safe in general.```

	$sql = "ALTER TABLE `tag` DROP COLUMN `is_hidden`";
 	$retval &= (Dba::write($sql) !== false);
 	print(sprintf("in update_370018 alter-table drop is_hidden retval=%s\n", $retval)); flush(); ob_flush();
        $sql = "ALTER TABLE `tag` ADD COLUMN `is_hidden` TINYINT(1) NOT NULL DEFAULT 0";
        $retval &= (Dba::write($sql) !== false);
	print(sprintf("in update_370018 alter table add is_hidden retval=%s\n", $retval)); flush(); ob_flush();
        return $retval;
```

Now I'm up to 550005!

**Reply by sgtell** [reply]
Actually, 550005 worked, and now I can log in and see the music catalogs.    All is not quite right, however:
 - most (but not all) artwork is missing
 - browse by "album artist" is completely blank.  seems like this might be a database thing.
It does look like this is getting usable, however.  
Thanks for the help!
One experiment i might try is re-loading the old saved mysqldump, and seeing if the updates run all the way through without stopping now, and how that turns out.  
If you want my mysql database as a testcase, lets figure out how to communicate privately.

**Reply by lachlan-00** [reply]
That's awesome.

I would do a catalog add to make sure all the maps and things get added. there is post-processing that should be done but may be missing for you as it's been behind from the site perspective. that should cover most missing things in the site though that you can't see/browse

For images i know there were a lot of table changes so that might have not copied or dropped before the errors? They would probably have been lost from your current DB so a restore would help.

I'm really glad that's gone through and i can fix that column as well with another check and we can work out where your image table went awry.

If you want to my gmail is public on my profile? (it's all through this repo) or use the security emails and https://github.com/ampache/ampache/security/policy

Feel free to take out any data you want with dummy data like names or emails and if you want a place to upload i can provide a nextcloud folder via email.

**Reply by lachlan-00** [reply]
added check for is_hidden which may just be from repeated errors but we should do this when checking a table/column really

```

    /*
     * update 370018
     *
     * Enhance tag persistent merge reference.
     */
    public static function update_370018(): bool
    {
        $retval  = true;
        $sql     = "DROP TABLE IF EXISTS `tag_merge`;";
        $retval &= (Dba::write($sql) !== false);
        $sql = "CREATE TABLE IF NOT EXISTS `tag_merge` (`tag_id` int(11) NOT NULL, `merged_to` int(11) NOT NULL, PRIMARY KEY (`tag_id`,`merged_to`), KEY `merged_to` (`merged_to`)) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;";
        $retval &= (Dba::write($sql) !== false);
        $sql        = "DESCRIBE `tag`";
        $db_results = Dba::read($sql);
        $is_hidden  = false;
        while ($row = Dba::fetch_assoc($db_results)) {
            if ($row['Field'] == 'merged_to') {
                $sql = "INSERT INTO `tag_merge` (`tag_id`, `merged_to`) SELECT `tag`.`id`, `tag`.`merged_to` FROM `tag` WHERE `merged_to` IS NOT NULL";
                $retval &= (Dba::write($sql) !== false);
                if ($retval) {
                    // don't drop until you've confirmed a merge
                    $sql = "ALTER TABLE `tag` DROP COLUMN `merged_to`";
                    $retval &= (Dba::write($sql) !== false);
                }
            }
            if ($row['Field'] == 'is_hidden') {
                $is_hidden = true;
            }
        }
        if (!$is_hidden) {
            $sql = "ALTER TABLE `tag` ADD COLUMN `is_hidden` TINYINT(1) NOT NULL DEFAULT 0";
            $retval &= (Dba::write($sql) !== false);
        }

        return $retval;
    }
```

I think for image data it should already be in it's own table from `update_360003` so that should be solveable.

update_550005 is also very slow for some people and i need to make that simpler. it also should be fixing up the mapping tables but that doesn't seem right for album_artists. there may be an issue outside the db update to fix
